source/alloc/src/heap.rs

```diff
@@ -1,505 +1,128 @@
-
-            if used > size {
-                return Err(());
-            }
-
-            cursor = cursor.wrapping_add(stub_offset);
-            stub_ptr = cursor.cast::<Recycle>();
-            stub_ptr.write(Recycle {
-                links: Links::new_stub(),
-                node_layout: stub_layout,
-            });
-            cursor = cursor.add(stub_size);
-        }
+static OOM_WAITER: WaitQueue = WaitQueue::new();
+static INHIBIT_ALLOC: AtomicBool = AtomicBool::new(false);
 
-        // Next we allocate ourselves
-        {
-            let aheap_layout = Layout::new::<Self>();
-            let aheap_offset = cursor.align_offset(aheap_layout.align());
-            let aheap_size = aheap_layout.size();
-            used += aheap_offset;
-            used += aheap_size;
-
-            if used > size {
-                return Err(());
+pub async fn alloc(layout: Layout) -> NonNull<u8> {
+    loop {
+        unsafe {
+            match NonNull::new(alloc::alloc::alloc(layout.clone())) {
+                Some(nn) => return nn,
+                None => {
+                    let _ = OOM_WAITER.wait().await;
+                    continue;
+                }
             }
-
-            cursor = cursor.wrapping_add(aheap_offset);
-            aheap_ptr = cursor.cast::<Self>();
-
-            // Increment the cursor, as we use it for the heap initialization
-            cursor = cursor.add(aheap_size);
-
-            let heap = Heap::new(cursor, end - (cursor as usize));
-
-            aheap_ptr.write(Self {
-                freelist: MpscQueue::new_with_static_stub(&*stub_ptr),
-                state: AtomicU8::new(Self::BUSY_LOCKED),
-                heap: UnsafeCell::new(heap),
-                heap_wait: WaitQueue::new(),
-                inhibit_alloc: AtomicBool::new(false),
-                any_frees: AtomicBool::new(false),
-            });
         }
-
-        // Everything else is now our allocation space.
-        let aheap = NonNull::new_unchecked(aheap_ptr);
-        let aheap_ref: &'static AHeap = aheap.as_ref();
-
-        // Creating exclusive access to the inner heap is acceptable, as we
-        // have marked ourselves with "BUSY_LOCKED", acting as a mutex.
-        let guard = HeapGuard { aheap: aheap_ref };
-
-        // Well that went great, I think!
-        Ok((aheap, guard))
     }
+}
 
-    pub(crate) unsafe fn release_node(&'static self, node: NonNull<Recycle>) {
-        // Can we immediately lock the allocator, avoiding the free list?
-        if let Ok(mut guard) = self.lock() {
-            let layout: Layout = (*node.as_ptr()).node_layout;
-            guard.get_heap().deallocate(node.cast::<u8>(), layout);
-            return;
-        }
-
-        // Nope! Stick it in the free list
-        let node_ref = NodeRef { node };
-        self.freelist.enqueue(node_ref);
-    }
+#[inline(always)]
+pub unsafe fn dealloc(ptr: *mut u8, layout: Layout) {
+    alloc::alloc::dealloc(ptr, layout)
+}
```

Comment:
nit, take it or leave it (not important): can't this just be 
```suggestion
pub use alloc::alloc::dealloc;
```